src/cluster/placement/selector/mirrored.go

```diff
@@ -416,16 +416,30 @@ func groupHostsWithIsolationGroupCheck(hosts []host, rf int) (groups [][]host, u
 	return groups, ungrouped
 }
 
-func groupInstancesByHostPort(hostGroups [][]host) ([][]placement.Instance, error) {
+func groupInstancesByHostPort(hostGroups [][]host, skipPortMatching bool) ([][]placement.Instance, error) {
 	var instanceGroups = make([][]placement.Instance, 0, len(hostGroups))
 	for _, hostGroup := range hostGroups {
 		for port, instance := range hostGroup[0].portToInstance {
 			instanceGroup := make([]placement.Instance, 0, len(hostGroup))
 			instanceGroup = append(instanceGroup, instance)
 			for _, otherHost := range hostGroup[1:] {
-				otherInstance, ok := otherHost.portToInstance[port]
-				if !ok {
-					return nil, fmt.Errorf("could not find port %d on host %s", port, otherHost.name)
+				var (
+					otherInstance placement.Instance
+					ok            bool
+				)
+				if !skipPortMatching {
+					otherInstance, ok = otherHost.portToInstance[port]
+					if !ok {
+						return nil, fmt.Errorf("could not find port %d on host %s", port, otherHost.name)
+					}
+				} else {
+					if len(otherHost.portToInstance) == 0 {
+						return nil, fmt.Errorf("could not find any available instance on host %s", otherHost.name)
+					}
+					for _, v := range otherHost.portToInstance {
```

Comment:
Hmm, I don't quite get the for-loop. Why do we only grab the first instance in this list?